php-checks/src/test/resources/checks/SelfKeywordUsageCheck.php

```diff
@@ -23,4 +25,35 @@ public static function f2() {
     return static::$field;     // OK
   }
 
+  public static final function f3() {
+  }
+
+}
+
+final class B {
+  public static $field;
+
+  public static function f1() {
+    self::f2();                // OK, class is final
+    return self::$field;       // OK, class is final
+  }
+
+  public static function f2() {}
+}
+
+class D {
+  public static function f1() {
+    self::f2();      // NOK
+    class C {
```

Comment:
@vilchik-elena I don't think that classes can be nested, but our grammar allows it...